### PR TITLE
Added CornerRadius property To Frame control

### PR DESCRIPTION
**Pull request by andreinitescu**
### Description of Change

Added CornerRadius property to the Frame control to customize Frame's corner radius value.
### Bugs Fixed

N/A
### API Changes

Added:
- float Frame.CornerRadius { get; set; } //Bindable Property
### Behavioral Changes

The property default value ensures the appearance of the Frame control in current apps remains unchanged.
### PR Checklist
- [ ] Has tests (if omitted, state reason in description)
- [ ] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [ ] Consolidate commits as makes sense

### CODE REVIEW

**Comment by andreinitescu** [comment]
Sorry, I couldn't find any info on what updating "docs" is.
Is it just updating this file https://github.com/xamarin/Xamarin.Forms/blob/master/docs/Xamarin.Forms.Core/Xamarin.Forms/Frame.xml ?

**Comment by adrianknight89** [comment]
This reminded me of https://forums.xamarin.com/discussion/comment/224717.

**Comment by andreinitescu** [comment]
@adrianknight89 if that's really a bug, not sure if it's related to this PR

**Comment by andreinitescu** [comment]
if someone can share to me how to update docs...

**Comment by adrianknight89** [comment]
You need to look at `Frame.xml` in APIDocs -> Xamarin.Forms.Core -> Xamarin.Forms. If Visual Studio shows the folder blank, select the folder and click "Show All Files" at the top toolbar. Then, right click and include `Frame.xml` in your project.

You could use `OutlineColor` as your base framework. There is a member for both `OutlineColor` and `OutlineColorProperty`. You should include something similar. In terms of `AssemblyVersion`, I think it needs to be `<AssemblyVersion>2.0.0.0</AssemblyVersion>`.

@rmarinho Should the assembly version show 2.3?

There is a third reference to `OutlineColor` at the top. I think that's so it shows at the top of the property window when intellisense appears with a dot character as in `myFrame.` You probably need an entry there as well. Again @rmarinho can confirm.

File: Xamarin.Forms.Core/Frame.cs [hunk 1]
```diff
@@ -11,6 +11,8 @@ public class Frame : ContentView, IElementConfiguration<Frame>
 
 		public static readonly BindableProperty HasShadowProperty = BindableProperty.Create("HasShadow", typeof(bool), typeof(Frame), true);
 
+		public static readonly BindableProperty CornerRadiusProperty = BindableProperty.Create(nameof(CornerRadius), typeof(float), typeof(Frame), -1.0f);
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Frame.cs]: what's the meaning of "-1" ?
- don't use a corner radius, or
- use whatever default on the platform

File: Xamarin.Forms.Platform.Android/AppCompat/FrameRenderer.cs [hunk 2]
```diff
@@ -286,5 +290,22 @@ void UpdateShadow()
...
+		void UpdateCornerRadius()
+		{
+			if (_defaultCornerRadius == -1f)
+			{
+				_defaultCornerRadius = Radius;
+			}
+
+			float cornerRadius = Element.CornerRadius;
+
+			if (cornerRadius == -1f)
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Platform.Android/AppCompat/FrameRenderer.cs]: I'd write `if (cornerRadius < 0)`, so you don't have to think about what happens if the user sets a value of, say, `-2`

File: Xamarin.Forms.Core/Frame.cs (hunk 1 above)
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Frame.cs]: doesn't this require a validate delegate ?

File: Xamarin.Forms.Platform.WP8/FrameRenderer.cs [hunk 3]
```diff
@@ -53,5 +55,15 @@ void UpdateBorder()
 			else
 				Control.BorderBrush = new Color(0, 0, 0, 0).ToBrush();
 		}
+
+		void UpdateCornerRadius()
+		{
+			float cornerRadius = Element.CornerRadius;
+
+			if (cornerRadius == -1f)
+				cornerRadius = 5f; // default corner radius
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Platform.WP8/FrameRenderer.cs]: where is this default defined ?

File: Xamarin.Forms.Platform.iOS/Renderers/FrameRenderer.cs [hunk 4]
```diff
@@ -25,7 +25,13 @@ protected override void OnElementPropertyChanged(object sender, PropertyChangedE
 
 		void SetupLayer()
 		{
-			Layer.CornerRadius = 5;
+			float cornerRadius = Element.CornerRadius;
+
+			if (cornerRadius == -1f)
+				cornerRadius = 5f; // default corner radius
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Platform.iOS/Renderers/FrameRenderer.cs]: same. I like to know where magic numbers are coming from

File: Xamarin.Forms.Core/Frame.cs (hunk 1 above)
Comment (andreinitescu) [inline Xamarin.Forms.Core/Frame.cs]: Per "Behavioral Changes", "-1" is the default value and it assures backward compatibility with existing apps. The previous implementation used different values on each platform.
I don't think "-1" can be interpreted as "don't use a corner radius". Being a corner radius, -1 seems like an invalid value anyway. It should make sense that "0" value is for no corner radius. 

To have backward compatibility, my options for default value were a nullable float and -1. I always preferred nullable, I used it for a long time, but for some reason, seeing recently other people's code, I started to think about using magic consts. And I don't like magic consts :) They need an extra constant field. On the other hand, nullable types are a little bit harder to use. Maybe I'm overthinking. 

Looking forward to hear your thoughts..

File: Xamarin.Forms.Platform.Android/AppCompat/FrameRenderer.cs (hunk 2 above)
Comment (andreinitescu) [inline Xamarin.Forms.Platform.Android/AppCompat/FrameRenderer.cs]: I was thinking negative values do not make sense anyway. Beside -1, if someone uses it, then unexpected behavior can happen.

File: Xamarin.Forms.Platform.iOS/Renderers/FrameRenderer.cs (hunk 4 above)
Comment (andreinitescu) [inline Xamarin.Forms.Platform.iOS/Renderers/FrameRenderer.cs]: please see my previous answer.

File: Xamarin.Forms.Platform.WP8/FrameRenderer.cs (hunk 3 above)
Comment (andreinitescu) [inline Xamarin.Forms.Platform.WP8/FrameRenderer.cs]: I hate hardcoded magic numbers too, but unfortunately this is how it was previously, see previous code. And I thought I shouldn't add more changes to the code.

File: Xamarin.Forms.Core/Frame.cs (hunk 1 above)
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Frame.cs]: `-1` is fine as an internal value. API consumer do not have to know about it's value, they can reset it to default using `frame.ClearValue(CornerRadiusProperty)`

File: Xamarin.Forms.Platform.WP8/FrameRenderer.cs (hunk 3 above)
Comment (StephaneDelcroix) [inline Xamarin.Forms.Platform.WP8/FrameRenderer.cs]: ok

File: Xamarin.Forms.Platform.Android/AppCompat/FrameRenderer.cs (hunk 2 above)
Comment (andreinitescu) [inline Xamarin.Forms.Platform.Android/AppCompat/FrameRenderer.cs]: @StephaneDelcroix I just checked, on Android for example a negative value doesn't produce any error, the corners are drawn like with 0 radius. It's up to the platform to interpret it.

File: Xamarin.Forms.Platform.iOS/Renderers/FrameRenderer.cs (hunk 4 above)
Comment (activa) [inline Xamarin.Forms.Platform.iOS/Renderers/FrameRenderer.cs]: Equality comparison between floating point numbers is not a good idea anyway. It should be changed to **if (cornerRadius < 0f)** IMHO.

**Comment by andreinitescu** [comment]
I still don't know what it needs to be changed, if anything. Also some follow up on is required on some items I replied.

File: Xamarin.Forms.Platform.iOS/Renderers/FrameRenderer.cs (hunk 4 above)
Comment (StephaneDelcroix) [inline Xamarin.Forms.Platform.iOS/Renderers/FrameRenderer.cs]: @activa Equality comparison is guaranteed to work for floating point numbers when the numbers are power of two. This is not an issue here.

Comment (activa) [inline Xamarin.Forms.Platform.iOS/Renderers/FrameRenderer.cs]: It's not an issue in this particular case, but it's still a bad idea in general. Any equality comparison between floating point numbers should raise a red flag.

Comment (andreinitescu) [inline Xamarin.Forms.Platform.iOS/Renderers/FrameRenderer.cs]: I know that, but it's really not an issue in this case. As stated before, the reason I didn't want to just use "smaller than" is to specifically show that I am testing for the default value. A better way would probably have been to test for `CornerRadiusProperty.DefaultValue`

**Comment by StephaneDelcroix** [comment]
OK. I re-reviewed this. I'm fine with the `-1f`. I still want a validateValue delegate that checks if the value is -1 or >= 0.

Do you think you can do that?

For the docs update, there's a script to run at the top level of the project (update-docs.ps1 iirc), and then you commit the parts that are meaningful. You don't have to actually write the doc, only the stubs.

**Comment by andreinitescu** [comment]
@StephaneDelcroix I will work on that.

**Comment by ahmedalejo** [comment]
@andreinitescu instead of checking for default value why don´t you simply set `CornerRadiusProperty.DefaultValue` on the Bindable Property to `5f`? This way there won´t be these questioning. In the docs will state "_..._ defaults to `5f` _..._"

**Your changes will be very useful**. Especially for those who needed to implement badges. I normally use buttons for badges but lack cross-platform corner radius and flexibility of it´s contents

**Comment by andreinitescu** [comment]
@ahmedalejo That's a good question. 
Indeed all renderers use `5` as default value, except the Android AppCompat renderer which unfortunately computes the default value differently:
https://github.com/andreinitescu/Xamarin.Forms/blob/3a4f123774d933be8d67328b9aaf2d3f75350937/Xamarin.Forms.Platform.Android/AppCompat/FrameRenderer.cs#L298
the default value is the unmodified value of the `CardView` `Radius`.

Based on my tests in an app, if I use a `5` as default value, the pixel value is much different than the unmodified value of the `CardView` `Radius`:

    public class MainActivity : global::Xamarin.Forms.Platform.Android.FormsAppCompatActivity
    {
        protected override void OnCreate(Bundle bundle)
        {
            TabLayoutResource = Resource.Layout.Tabbar;
            ToolbarResource = Resource.Layout.Toolbar;

            base.OnCreate(bundle);

            CardView cv = new CardView(this);  // cv.Radius is 7.2
            var r = Xamarin.Forms.Platform.Android.ContextExtensions.ToPixels(this, 5); // r is 18

Remember we have to make sure adding this new property doesn't modify the appearance of the existing apps, therefore we can't modify the default value.

Or maybe I'm missing something?

**Comment by StephaneDelcroix** [comment]
@andreinitescu I'll take it from here, it'll be simpler for me to just do the changes instead of having back and forth about detail. Does that works for you ?

**Comment by andreinitescu** [comment]
@StephaneDelcroix I'm really not sure a validate delegate is the best way to go in this situation.
Why add further computation in the binding? Here and there, as insignificant it might look, these tiny things add up. And once it's implemented, it's hard to go over all code and try squeeze it back for better performance. We should really try to avoid it whenever possible in my opinion.

If using a nullable `int?` for the property value is a better way, then maybe use that instead?

**Comment by andreinitescu** [comment]
@StephaneDelcroix Discussions are useful, but if it's too time consuming for you sure, go ahead please, I feel better if I know it leaves you time for other things to do in XF, there's so many.
But please try to take into account my suggestion from previous comment.

**Comment by StephaneDelcroix** [comment]
@andreinitescu the cost/benefit ratio of the validate value delegate is so small I don't even know why we discuss it :)

Also, please continue the discussion on #496

**Comment by andreinitescu** [comment]
@StephaneDelcroix yeah I know it's insignificant, but add that to the code in the `UpdateCornerRadius` and all the other stuff which is run when control is initialized, multiply it by all controls you use in a real app, and it doesn't look pretty. 
It's worth discussing on this, but I don't see the utility for validating the corner radius. On native, there's no error if you put negative values for corner radius, I don't understand why XF has do to it.

**Pull request closed without merging**